File: courses/codes/optional-HW-1-CCC/make-your-move.py
source = input().split(' ')
x0 = int(source[0])
y0 = int(source[1])

# initialization
x, y = x0, y0
is_on = False
bat = 50

print(x0, y0)

cmd = input()

while cmd != 'Q':

    if cmd == 'C':
        bat = 100
    elif cmd == 'P' and bat > 0:
        is_on = not is_on
    else:
        arr = cmd.split(' ')
        dirn = arr[0]
        steps = 0 if not is_on else (1 if len(arr) == 1 else int(arr[1]))

        if dirn == 'F':
            y += min(steps, bat)
        elif dirn == 'B':
            y -= min(steps, bat)
        elif dirn == 'L':
            x -= min(steps, bat)
        else:
            x += min(steps, bat)

        bat -= min(steps, bat)

    if bat == 0:
        is_on = False

    print(x, y)

    cmd = input()

d = ((x - x0) ** 2 + (y - y0) ** 2) ** 0.5
print("%.2f" % d)

File: courses/codes/optional-HW-1-CCC/make-your-move.cpp
#include <iostream>
#include <string>
#include <sstream>
#include <cmath>

int main() {
    std::string input;
    
    // Read the initial position
    std::getline(std::cin, input);
    std::istringstream iss(input);
    int x0, y0;
    iss >> x0 >> y0;
    
    // Initializations
    int x = x0;
    int y = y0;
    bool isOn = false;
    int bat = 50;
    
    std::cout << x << " " << y << std::endl;
    
    // Read commands
    while (std::getline(std::cin, input) && input != "Q") {
        
        if (input == "C") {
            bat = 100;
        } else if (input == "P" && bat > 0) {
            isOn = !isOn;
        } else {
            std::istringstream cmdStream(input);
            std::string dirn;
            cmdStream >> dirn;
            int steps;
            
            if (isOn) {
                if (!(cmdStream >> steps)) { // If steps are not provided, assume 1 step
                    steps = 1;
                }
            } else {
                steps = 0;
            }
            
            int moveSteps = std::min(steps, bat);
            
            if (dirn == "F") {
                y += moveSteps;
            } else if (dirn == "B") {
                y -= moveSteps;
            } else if (dirn == "L") {
                x -= moveSteps;
            } else if (dirn == "R") {
                x += moveSteps;
            }
            
            bat -= moveSteps;
        }
        
        if (bat == 0) {
            isOn = false;
        }
        
        std::cout << x << " " << y << std::endl;
    }
    
    double d = std::sqrt((x - x0) * (x - x0) + (y - y0) * (y - y0));
    std::cout << std::fixed;
    std::cout.precision(2);
    std::cout << d << std::endl;
    
    return 0;
}